File: bindings/python/src/graphqlite/graph/edges.py
"""Edge operations mixin for Graph class."""

from typing import Any, Optional

from ._base import BaseMixin
from ..utils import sanitize_rel_type


class EdgesMixin(BaseMixin):
    """Mixin providing edge CRUD operations."""

    def has_edge(self, source_id: str, target_id: str, rel_type: Optional[str] = None) -> bool:
        """
        Check if an edge exists between two nodes.

        Args:
            source_id: Source node id
            target_id: Target node id
            rel_type: Optional relationship type to check for

        Returns:
            True if edge exists, False otherwise
        """
        rel_pattern = f":{sanitize_rel_type(rel_type)}" if rel_type else ""
        result = self._conn.cypher(
            f"MATCH (a {{id: $src}})-[r{rel_pattern}]->"
            f"(b {{id: $tgt}}) "
            f"RETURN count(r) AS cnt",
            params={"src": source_id, "tgt": target_id},
        )
        if len(result) == 0:
            return False
        cnt = result[0].get("cnt", 0)
        return int(cnt) > 0 if cnt else False

    def get_edge(self, source_id: str, target_id: str, rel_type: Optional[str] = None) -> Optional[dict]:
        """
        Get edge properties between two nodes.

        Args:
            source_id: Source node id
            target_id: Target node id
            rel_type: Optional relationship type to retrieve

        Returns:
            Edge dict or None if not found
        """
        rel_pattern = f":{sanitize_rel_type(rel_type)}" if rel_type else ""
        result = self._conn.cypher(
            f"MATCH (a {{id: $src}})-[r{rel_pattern}]->"
            f"(b {{id: $tgt}}) RETURN r",
            params={"src": source_id, "tgt": target_id},
        )
        if len(result) == 0:
            return None
        return result[0].get("r")

    def upsert_edge(
        self,
        source_id: str,
        target_id: str,
        edge_data: dict[str, Any],
        rel_type: str = "RELATED"
    ) -> None:
        """
        Create or update an edge between two nodes.

        If an edge of the same type already exists, its properties are updated
        (merge semantics -- existing properties not in edge_data are preserved).
        If no edge of that type exists, a new one is created.
        Both source and target nodes must exist.

        Args:
            source_id: Source node id
            target_id: Target node id
            edge_data: Dictionary of edge properties
            rel_type: Relationship type label
        """
        safe_rel_type = sanitize_rel_type(rel_type)

        self._conn.cypher(
            f"MATCH (a {{id: $src}}), (b {{id: $tgt}}) "
            f"MERGE (a)-[r:{safe_rel_type}]->(b)",
            params={"src": source_id, "tgt": target_id},
        )

        if edge_data:
            params = {"src": source_id, "tgt": target_id}
            set_parts = []
            for i, (k, v) in enumerate(edge_data.items()):
                param_name = f"v{i}"
                set_parts.append(f"r.{k} = ${param_name}")
                params[param_name] = v
            set_str = ", ".join(set_parts)
            self._conn.cypher(
                f"MATCH (a {{id: $src}})-[r:{safe_rel_type}]->"
                f"(b {{id: $tgt}}) SET {set_str}",
                params=params,
            )

    def delete_edge(self, source_id: str, target_id: str, rel_type: Optional[str] = None) -> None:
        """
        Delete edge between two nodes.

        Args:
            source_id: Source node id
            target_id: Target node id
            rel_type: Optional relationship type to delete
        """
        rel_pattern = f":{sanitize_rel_type(rel_type)}" if rel_type else ""
        self._conn.cypher(
            f"MATCH (a {{id: $src}})-[r{rel_pattern}]->"
            f"(b {{id: $tgt}}) DELETE r",
            params={"src": source_id, "tgt": target_id},
        )

    def get_all_edges(self) -> list[dict]:
        """
        Get all edges with source and target info.

        Returns:
            List of dicts with 'source', 'target', and edge properties
        """
        result = self._conn.cypher(
            "MATCH (a)-[r]->(b) RETURN a.id AS source, b.id AS target, r"
        )
        return result.to_list()

File: bindings/rust/src/graph/edges.rs
//! Edge operations for Graph.

use super::Graph;
use crate::utils::{rel_type_pattern, sanitize_rel_type, PropertyValue};
use crate::{CypherResult, Result, Value};

impl Graph {
    /// Check if a directed edge exists from source to target.
    pub fn has_edge(&self, source_id: &str, target_id: &str, rel_type: Option<&str>) -> Result<bool> {
        let rel_pattern = rel_type_pattern(rel_type);
        let query = format!(
            "MATCH (a {{id: $src}})-[r{}]->(b {{id: $tgt}}) RETURN count(r) AS cnt",
            rel_pattern
        );
        let result = self.connection()
            .cypher_builder(&query)
            .param("src", source_id)
            .param("tgt", target_id)
            .run()?;
        if result.is_empty() {
            return Ok(false);
        }
        let cnt: i64 = result[0].get("cnt").unwrap_or(0);
        Ok(cnt > 0)
    }

    /// Get edge properties between two nodes.
    pub fn get_edge(&self, source_id: &str, target_id: &str, rel_type: Option<&str>) -> Result<Option<Value>> {
        let rel_pattern = rel_type_pattern(rel_type);
        let query = format!(
            "MATCH (a {{id: $src}})-[r{}]->(b {{id: $tgt}}) RETURN r",
            rel_pattern
        );
        let result = self.connection()
            .cypher_builder(&query)
            .param("src", source_id)
            .param("tgt", target_id)
            .run()?;
        if result.is_empty() {
            return Ok(None);
        }
        Ok(result[0].get_value("r").cloned())
    }

    /// Create or update an edge between two nodes.
    ///
    /// If an edge of the same type already exists, its properties are updated
    /// (merge semantics — existing properties not in `props` are preserved).
    /// If no edge of that type exists, a new one is created.
    pub fn upsert_edge<I, K, V>(
        &self,
        source_id: &str,
        target_id: &str,
        props: I,
        rel_type: &str,
    ) -> Result<()>
    where
        I: IntoIterator<Item = (K, V)>,
        K: AsRef<str>,
        V: Into<PropertyValue>,
    {
        let safe_rel_type = sanitize_rel_type(rel_type);

        let props: Vec<(String, PropertyValue)> = props
            .into_iter()
            .map(|(k, v)| (k.as_ref().to_string(), v.into()))
            .collect();

        let merge_query = format!(
            "MATCH (a {{id: $src}}), (b {{id: $tgt}}) MERGE (a)-[r:{}]->(b)",
            safe_rel_type
        );
        self.connection()
            .cypher_builder(&merge_query)
            .param("src", source_id)
            .param("tgt", target_id)
            .run()?;

        if !props.is_empty() {
            let set_parts: Vec<String> = props
                .iter()
                .map(|(k, v)| format!("r.{} = {}", k, v.to_cypher()))
                .collect();
            let set_str = set_parts.join(", ");
            let set_query = format!(
                "MATCH (a {{id: $src}})-[r:{}]->(b {{id: $tgt}}) SET {}",
                safe_rel_type, set_str
            );
            self.connection()
                .cypher_builder(&set_query)
                .param("src", source_id)
                .param("tgt", target_id)
                .run()?;
        }

        Ok(())
    }

    /// Delete the directed edge between two nodes.
    pub fn delete_edge(&self, source_id: &str, target_id: &str, rel_type: Option<&str>) -> Result<()> {
        let rel_pattern = rel_type_pattern(rel_type);
        let query = format!(
            "MATCH (a {{id: $src}})-[r{}]->(b {{id: $tgt}}) DELETE r",
            rel_pattern
        );
        self.connection()
            .cypher_builder(&query)
            .param("src", source_id)
            .param("tgt", target_id)
            .run()?;
        Ok(())
    }

    /// Get all edges in the graph.
    pub fn get_all_edges(&self) -> Result<CypherResult> {
        self.connection()
            .cypher("MATCH (a)-[r]->(b) RETURN a.id AS source, b.id AS target, r")
    }
}
